cassiopeia/solver/SpectralNeighborJoiningSolver.py

```diff
@@ -0,0 +1,281 @@
+    """
+
+    def __init__(
+        self,
+        dissimilarity_function: Optional[
+            Callable[
+                [np.ndarray, np.ndarray, int, Dict[int, Dict[int, float]]], float
+            ]
+        ] = dissimilarity_functions.weighted_hamming_distance,
+        add_root: bool = False,
+        prior_transformation: str = "negative_log",
+    ):
+        super().__init__(dissimilarity_function, add_root=add_root, prior_transformation=prior_transformation) #type: ignore
+
+    def get_dissimilarity_map(
+        self, 
+        cassiopeia_tree: CassiopeiaTree,
+        layer: Optional[str] = None
+    ) -> pd.DataFrame: 
+        """Outputs the first lambda matrix, where every subset is an individual node.
+
+        Args:
+            cassiopeia_tree: the CassiopeiaTree object passed into solve()
+            layer: Layer storing the character matrix for solving. If None, the
+                default character matrix is used in the CassiopeiaTree.
+
+        Returns:
+            DataFrame object of the lamba matrix, but similar in structure to DistanceSolver's dissimilarity_map.
+        """
+
+        # get similarity map and save it as private instance variable
+        self.__similarity_map = super().get_dissimilarity_map(cassiopeia_tree, layer)
+
+        # generate first (pairwise) lambda matrix
+        N = self.__similarity_map.shape[0]
+        self.lambda_indices: List[List[int]] = [[i] for i in range(N)]
+        node_names: np.ndarray = self.__similarity_map.index.values
+
+        lambda_matrix_arr = self.compute_lambda(self.lambda_indices)
+
+        lambda_matrix_df = pd.DataFrame(
+            lambda_matrix_arr, 
+            index=node_names,
+            columns=node_names
+        )        
+
+        return lambda_matrix_df
+
+
+    def find_cherry(self, dissimilarity_map: np.ndarray) -> Tuple[int, int]:
+        """Finds a pair of samples to join into a cherry.
+
+        With dissimilarity_map being the lambda matrix, this method finds the argmin pair of subsets of the lambda matrix. 
+
+        Args:
+            dissimilarity_matrix: Lambda matrix
+
+        Returns:
+            A tuple of intgers representing rows in the dissimilarity matrix
+                to join.
+        """
+
+        return np.unravel_index(np.argmin(dissimilarity_map, axis=None), dissimilarity_map.shape)
+
+
+    def compute_lambda(self, lambda_indices: List[List[int]]) -> np.ndarray:
+        """Computes the lambda matrix, whose entries consist of the second SVD of the R^A affinity matrix for every pair of nodes.
+
+        Computes the lambda(i,j) = SVD # 2 for R^(Ai U Aj) where i and j are the susbet indices in lambda_indices.
+
+        Args:
+            dissimilarity_map: A sample x sample dissimilarity map
+
+        Returns:
+            A matrix storing the Q-criterion for every pair of samples.
```

Comment:
This is not going to store the "Q-criterion" for every pair of samples, but rather the second singular value for every pair of clades.